syntax = "proto3";

package iart.indexer;

import "data.proto";

message Collection {
  string id = 1;
  string name = 2;
  bool is_public = 3;
}

message RoI {
  float x = 1;
  float y = 2;
  float width = 3;
  float height = 4;
}

message ValueField {
  string key = 1;

  oneof value {
    string string_val = 2;
    int32 int_val = 3;
    float float_val = 4;
  }
}

message Image {
  string id = 1;
  oneof image {
    string path = 2;
    bytes encoded = 3;
  }
  RoI roi = 4;
  repeated ValueField meta = 5;
  repeated ValueField origin = 6;

  Collection collection = 7;
}

message PluginConfig {
  string name = 1;
  string default = 2;
  string type = 3;
  string state = 4;
}

message PluginResult {
  string plugin = 1;
  string type = 2;
  string version = 3;
  PluginData result = 4;
}

enum DataType {
  UNKOWN_TYPE = 0;
  STRING_TYPE = 1;
  INT_TYPE = 2;
  FLOAT_TYPE = 3;
  BOOL_TYPE = 4;
  BYTES_TYPE = 5;
  IMAGE_TYPE = 6;
  BOUNDING_BOX_TYPE = 7;
}

message PluginParameter {
  string name = 1;
  bytes content = 2;
  repeated int32 shape = 3;
  DataType type = 4;
}

message PluginRun {
  string plugin = 1;
  repeated PluginParameter parameters = 2;
  repeated PluginData inputs = 3;
}